string_manipulation.rb

```diff
@@ -1,30 +1,127 @@
 # A method to reverse a string in place.
 def string_reverse(my_string)
-  puts "NOT IMPLEMENTED"
+  length = my_string.length
+  i = 0
+
+  if length >= 2
+  (length / 2).times do
+    temp = my_string[i]
+    my_string[i] = my_string[length - 1]
+    my_string[length - 1] = temp
+    i += 1
+    length -= 1
+    end
+  end
+  return my_string
 end
 
 # A method to reverse each word in a sentence, in place.
 def reverse_words(my_words)
-  puts "NOT IMPLEMENTED"
+  length = my_words.length
+
+  word_start = 0
+  word_end = 0
+
+  length.times do |i|
+    if my_words[i] == " " || i == length - 1
+
+      if my_words[i] == " "
+        word_end = i - 1
+      else
+        word_end = i
+      end
+
+      my_words[word_start..word_end] = string_reverse(my_words[word_start..word_end])
+      word_start = i + 1
+      word_end = i + 1
+    end
+    i += 1
+  end
+  return my_words
 end
 
 # A method to reverse the words in a sentence, in place.
 def reverse_sentence(my_sentence)
-  puts "NOT IMPLEMENTED"
+  my_sentence_reversed = string_reverse(my_sentence)
+  corrected_sentence = reverse_words(my_sentence_reversed)
+  return corrected_sentence
 end
 
 # A method to check if the input string is a palindrome.
 # Return true if the string is a palindrome. Return false otherwise.
 def palindrome_check(my_phrase)
-  puts "NOT IMPLEMENTED"
+  if my_phrase.length > 0
+    beg = 0
+    fin = my_phrase.length - 1
+
+    ((my_phrase.length) / 2).times do
+
+      while my_phrase[beg] == " "
+        beg += 1
+      end
+
+      while my_phrase[fin] == " "
+        fin -= 1
+      end
+
+      if my_phrase[beg] != my_phrase[fin]
+        return false
+      end
+    end
+  end
   return true
 end
 
+
 # A method that updates the string by replacing consecutive repeating characters
 # with a number representing the frequncy. The replacement is done only if the
 # string length will get reduced by the process.
 def encode_repeating(my_string)
-  puts "NOT IMPLEMENTED"
+  counter = 1
+  start = 0
+  finish = 0
+  i = 0
+
+  until i > my_string.length do
+
+    if i + 1 == my_string.length && counter >= 2
+      my_string[start] = my_string[start] + counter.to_s
+      i += 1
+      break
+    elsif (my_string[i] != my_string[i + 1]) && counter <= 2
+      start = i + 1
+      finish = i + 1
+      counter = 1
+      i += 1
+      next
+    elsif
+      (my_string[i] != my_string[i + 1]) && counter >= 3
+      my_string[start] = my_string[start] + counter.to_s

```

Comment:
You are actually inserting new strings in between here (i.e. increasing the length of my_string) and hence adding more space. To do this in a restricted array of characters, you'd do something like this:
`my_string[start] = my_string[i]`
`my_string[start+1] = counter.to_s`
`start += 2`
`i += 1`
`counter = 1`